// remotion/dist/cjs/resolve-video-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveVideoConfig = void 0;
const input_props_js_1 = require("./config/input-props.js");
const get_environment_js_1 = require("./get-environment.js");
const validate_dimensions_js_1 = require("./validation/validate-dimensions.js");
const validate_duration_in_frames_js_1 = require("./validation/validate-duration-in-frames.js");
const validate_fps_js_1 = require("./validation/validate-fps.js");
const resolveVideoConfig = ({ composition, editorProps: editorPropsOrUndefined, signal, }) => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const calculatedProm = composition.calculateMetadata
        ? composition.calculateMetadata({
            defaultProps: (_a = composition.defaultProps) !== null && _a !== void 0 ? _a : {},
            props: {
                ...((_b = composition.defaultProps) !== null && _b !== void 0 ? _b : {}),
                ...(editorPropsOrUndefined !== null && editorPropsOrUndefined !== void 0 ? editorPropsOrUndefined : {}),
                ...(typeof window === 'undefined' ||
                    (0, get_environment_js_1.getRemotionEnvironment)() === 'player-development' ||
                    (0, get_environment_js_1.getRemotionEnvironment)() === 'player-production'
                    ? {}
                    : (_c = (0, input_props_js_1.getInputProps)()) !== null && _c !== void 0 ? _c : {}),
            },
            abortSignal: signal,
        })
        : null;
    if (calculatedProm !== null &&
        typeof calculatedProm === 'object' &&
        'then' in calculatedProm) {
        return calculatedProm.then((c) => {
            var _a, _b, _c;
            const { height, width, durationInFrames, fps } = validateCalculated({
                calculated: c,
                composition,
            });
            return {
                width,
                height,
                fps,
                durationInFrames,
                id: composition.id,
                defaultProps: (_a = composition.defaultProps) !== null && _a !== void 0 ? _a : {},
                props: (_c = (_b = c.props) !== null && _b !== void 0 ? _b : composition.defaultProps) !== null && _c !== void 0 ? _c : {},
            };
        });
    }
    const data = validateCalculated({
        calculated: calculatedProm,
        composition,
    });
    if (calculatedProm === null) {
        return {
            ...data,
            id: composition.id,
            defaultProps: (_d = composition.defaultProps) !== null && _d !== void 0 ? _d : {},
            props: (_e = composition.defaultProps) !== null && _e !== void 0 ? _e : {},
        };
    }
    return {
        ...data,
        id: composition.id,
        defaultProps: (_f = composition.defaultProps) !== null && _f !== void 0 ? _f : {},
        props: (_h = (_g = calculatedProm.props) !== null && _g !== void 0 ? _g : composition.defaultProps) !== null && _h !== void 0 ? _h : {},
    };
};
exports.resolveVideoConfig = resolveVideoConfig;
const validateCalculated = ({ composition, calculated, }) => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const calculateMetadataErrorLocation = `calculated by calculateMetadata() for the composition "${composition.id}"`;
    const defaultErrorLocation = `of the "<Composition />" component with the id "${composition.id}"`;
    const width = (_b = (_a = calculated === null || calculated === void 0 ? void 0 : calculated.width) !== null && _a !== void 0 ? _a : composition.width) !== null && _b !== void 0 ? _b : undefined;
    (0, validate_dimensions_js_1.validateDimension)(width, 'width', (calculated === null || calculated === void 0 ? void 0 : calculated.width) ? calculateMetadataErrorLocation : defaultErrorLocation);
    const height = (_d = (_c = calculated === null || calculated === void 0 ? void 0 : calculated.height) !== null && _c !== void 0 ? _c : composition.height) !== null && _d !== void 0 ? _d : undefined;
    (0, validate_dimensions_js_1.validateDimension)(height, 'height', (calculated === null || calculated === void 0 ? void 0 : calculated.height) ? calculateMetadataErrorLocation : defaultErrorLocation);
    const fps = (_f = (_e = calculated === null || calculated === void 0 ? void 0 : calculated.fps) !== null && _e !== void 0 ? _e : composition.fps) !== null && _f !== void 0 ? _f : null;
    (0, validate_fps_js_1.validateFps)(fps, (calculated === null || calculated === void 0 ? void 0 : calculated.fps) ? calculateMetadataErrorLocation : defaultErrorLocation, false);
    const durationInFrames = (_h = (_g = calculated === null || calculated === void 0 ? void 0 : calculated.durationInFrames) !== null && _g !== void 0 ? _g : composition.durationInFrames) !== null && _h !== void 0 ? _h : null;
    (0, validate_duration_in_frames_js_1.validateDurationInFrames)(durationInFrames, {
        allowFloats: false,
        component: `of the "<Composition />" component with the id "${composition.id}"`,
    });
    return { width, height, fps, durationInFrames };
};
